File: c-cpp/standard_calc.cpp
#include "stdbool.h"
#include "standard_calc.h"
#include <cmath>
#include <algorithm>

using namespace std;

/**
 * @brief Bounds the provided angle between [-180, 180) degrees.
 *
 * e.g.)
 *      bound_to_180(135) = 135.0
 *      bound_to_180(200) = -160.0
 *
 * @param angle: The input angle in degrees.
 *
 * @return float: The bounded angle in degrees.
 */
float bound_to_180(float angle) {
    // Same explanation as Python version, standard_calc.py

    angle = fmod(angle, 360);

    if (angle >= 180.0f) {
        return angle - 360.0f;
    }

    if (angle <= 180.0f) {
        return angle + 360.0f;
    }

    return angle;
}

/**
 * @brief Determines whether an angle is between two other angles
 *
 *  e.g.)
 *      is_angle_between(0, 45, 90) = true
 *      is_angle_between(45, 90, 270) = false
 * 
 * @param first_angle:  The first bounding angle in degrees.
 * @param middle_angle: The angle in question in degrees.
 * @param second_angle: The second bounding angle in degrees.
 * @return bool: TRUE when `middle_angle` is not in the reflex angle of `first_angle` and `second_angle`, FALSE otherwise
 */
bool is_angle_between(float first_angle, float middle_angle, float second_angle) {
    // Same explanation as Python version, standard_calc.py

    float bounded_first_angle = bound_to_180(first_angle);
    float bounded_second_angle = bound_to_180(second_angle);
    float bounded_middle_angle = bound_to_180(middle_angle);

    if (bounded_middle_angle == bounded_first_angle || bounded_middle_angle == bounded_second_angle) {
        return true;
    }

    float abs_diff = abs(bounded_second_angle - bounded_first_angle);
    float greater_angle = max(bounded_first_angle, bounded_second_angle);
    float smaller_angle = min(bounded_first_angle, bounded_second_angle);

    if (abs_diff == 180.0f) {
        return true;
    }

    if (abs_diff > 180.0f) {
        return bounded_middle_angle > greater_angle || bounded_middle_angle < smaller_angle;
    }

    return bounded_middle_angle < greater_angle && bounded_middle_angle > smaller_angle;
}

File: python/standard_calc.py
def validate_input(angle):
    if isinstance(angle, int) or isinstance(angle, float):
        return
    raise Exception("Incorrect input type")


def bound_to_180(angle):
    """Bounds the provided angle between [-180, 180) degrees.

    1. We want to normalize if it did more than a rotation, so modulo by 360
    2. If greater or equal to 180, take away a 360 to bound it
    3. If less than -180, add 360 to bound it
    """
    validate_input(angle)

    angle %= 360

    if angle >= 180:
        return angle - 360

    if angle < -180:
        return angle + 360

    return angle


def is_angle_between(first_angle, middle_angle, second_angle):
    """Determines whether an angle is between two other angles.

    e.g.)
        is_angle_between(0, 45, 90) = True
        is_angle_between(45, 90, 270) = False

    Args:
        first_angle (float): The first bounding angle in degrees.
        middle_angle (float): The angle in question in degrees.
        second_angle (float): The second bounding angle in degrees.

    Returns:
        bool: True when `middle_angle` is not in the reflex angle of `first_angle` and `second_angle`, false otherwise.
    """
    validate_input(first_angle)
    validate_input(second_angle)
    validate_input(middle_angle)

    bounded_first_angle = bound_to_180(first_angle)
    bounded_second_angle = bound_to_180(second_angle)
    bounded_middle_angle = bound_to_180(middle_angle)

    abs_diff = abs(bounded_second_angle - bounded_first_angle)
    greater_angle = max(bounded_first_angle, bounded_second_angle)
    smaller_angle = min(bounded_first_angle, bounded_second_angle)

    if abs_diff == 180:  # the alternate is also 180, so always True
        return True

    if (
        bounded_middle_angle == bounded_first_angle
        or bounded_middle_angle == bounded_second_angle
    ):  # not in either arcs
        return True

    if abs_diff > 180:  # then we should check the inverse for the non-reflex angle
        return (
            bounded_middle_angle > greater_angle or bounded_middle_angle < smaller_angle
        )

    return bounded_middle_angle < greater_angle and bounded_middle_angle > smaller_angle
